src/herder/test/UpgradesTests.cpp

```diff
@@ -553,6 +624,164 @@ TEST_CASE("Ledger Manager applies upgrades properly", "[upgrades]")
+                SECTION("no updated entries")
+                {
+                    testInvalidXdr();
+                }
+                SECTION("duplicate entries")
+                {
+                    badConfigUpgradeSet.updatedEntry.emplace_back(
+                        CONFIG_SETTING_CONTRACT_MAX_SIZE_BYTES);
+                    badConfigUpgradeSet.updatedEntry.emplace_back(
+                        CONFIG_SETTING_CONTRACT_MAX_SIZE_BYTES);
+                    testInvalidXdr();
+                }
+            }
+        }
+        SECTION("bad value")
+        {
+            REQUIRE(Upgrades::isValidForApply(
+                        toUpgradeType(makeConfigUpgrade(
+                            *makeMaxContractSizeBytesTestUpgrade(ltx, 0))),
+                        outUpgrade, *app, ltx,
+                        header) == Upgrades::UpgradeValidity::INVALID);
+        }
+    }
+
+    SECTION("validate for nomination")
+    {
+        LedgerUpgradeType outUpgradeType;
+        SECTION("valid")
+        {
+            REQUIRE(Upgrades(scheduledUpgrades)
+                        .isValid(
+                            toUpgradeType(makeConfigUpgrade(*configUpgradeSet)),
+                            outUpgradeType, true, *app, ltx, header));
+        }
+        SECTION("not valid")
+        {
+            SECTION("no upgrade scheduled")
+            {
+                REQUIRE(!Upgrades().isValid(
+                    toUpgradeType(makeConfigUpgrade(*configUpgradeSet)),
+                    outUpgradeType, true, *app, ltx, header));
+            }
+            SECTION("inconsistent value")
+            {
+                REQUIRE(!Upgrades(scheduledUpgrades)
+                             .isValid(toUpgradeType(makeConfigUpgrade(
+                                          *makeMaxContractSizeBytesTestUpgrade(
+                                              ltx, 12345))),
+                                      outUpgradeType, true, *app, ltx, header));
+            }
+        }
+    }
+}
+
+TEST_CASE("config upgrades applied to ledger", "[upgrades]")
+{
+    VirtualClock clock;
+    auto cfg = getTestConfig(0);
+    cfg.TESTING_UPGRADE_LEDGER_PROTOCOL_VERSION =
+        static_cast<uint32_t>(SOROBAN_PROTOCOL_VERSION) - 1;
+    auto app = createTestApplication(clock, cfg);
+
+    // Need to actually execute the upgrade to v20 to get the config
+    // entries initialized.
+    executeUpgrade(*app, makeProtocolVersionUpgrade(
+                             static_cast<uint32_t>(SOROBAN_PROTOCOL_VERSION)));
+
+    SECTION("unknown config upgrade set results in exception")
+    {
+        // TODO: Use function
+        auto contractID = autocheck::generator<Hash>()(5);
+        auto upgradeHash = autocheck::generator<Hash>()(5);
+        auto ledgerUpgrade = LedgerUpgrade{LEDGER_UPGRADE_CONFIG};
+        ledgerUpgrade.newConfig() = ConfigUpgradeKey{contractID, upgradeHash};
+        REQUIRE_THROWS_AS(executeUpgrade(*app, ledgerUpgrade),
```

Comment:
NB: I believe exception is fine here as we shouldn't get to upgrade execution for the invalid/unknown upgrades. But we need to make sure we don't crash during validation itself.

Reply:
So this is still an open question I have. If a valid configUpgrade is nominated, is it possible so change the underlying ContractData entry that holds the ConfigUpgradeSet before the upgrade is applied? If so, that scenario needs to be handled.

Reply:
Hmm, yeah, it might change, so we need to verify that hash in the key matches the content hash. That can happen at validation time though, as upgrades happen after all the transactions. At apply time the upgrade set should be guaranteed to be correct and hence it should be ok to throw an exception (as that would be a bug)

Reply:
The hash is already verified in `ConfigUpgradeSetFrame::isValidXDR`.

Reply:
So we should be good, right? Also we load the actual entry just once anyway, so there shouldn't be mismatch between the config upgrade set at validation and application time.

Reply:
Added a section to ignore invalid upgrades.